### ISSUE TITLE
LoadError when using Unicorn and Nginx

### ISSUE THREAD

**Issue by jdraths** [GoogleCloudPlatform/google-cloud-ruby#1193]
Having a crash in *production environment only* on the `require 'google/cloud/vision'` line.
This all works fine in development.
What could I have wrong in production?

> /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `require': No such file to load -- google/gax (LoadError)

>...

>from /home/rails/app/services/cloud_vision.rb:5:in `<top (required)>'

**app/services/cloud_vision.rb**
```ruby
# app/services/cloud_vision.rb

### this is for [google-cloud-vision gem](http://googlecloudplatform.github.io/google-cloud-ruby/#/docs/google-cloud-vision/master/google/cloud/vision)

require 'google/cloud/vision'
require 'base64'
require 'googleauth'
require 'open-uri'
class CloudVision
	
    def initialize
    ### vision service
	@vision = Google::Cloud::Vision.new
    end
#...
end
```

I have `google-gax` in my **gemfile.lock**

```ruby
    google-cloud-core (0.21.1)
      googleauth (~> 0.5.1)
    google-cloud-vision (0.21.1)
      google-cloud-core (~> 0.21.0)
      google-gax (~> 0.6.0)
      google-protobuf (~> 3.0)
      googleapis-common-protos (~> 1.3)
      grpc (~> 1.0)
    google-gax (0.6.0)
      googleapis-common-protos (~> 1.3.1)
      googleauth (~> 0.5.1)
      grpc (~> 1.0)
      rly (~> 0.2.3)
    google-protobuf (3.0.2)
    googleapis-common-protos (1.3.4)
      google-protobuf (~> 3.0)
      grpc (~> 1.0)
    googleauth (0.5.1)
      faraday (~> 0.9)
      jwt (~> 1.4)
      logging (~> 2.0)
      memoist (~> 0.12)
      multi_json (~> 1.11)
      os (~> 0.9)
      signet (~> 0.7)
```

**stack trace**
```
/usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `require': No such file to load -- google/gax (LoadError)
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `block in require'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:240:in `load_dependency'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `require'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/google-cloud-vision-0.21.1/lib/google/cloud/vision/v1/image_annotator_api.rb:28:in `<top (required)>'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `require'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `block in require'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:240:in `load_dependency'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `require'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/google-cloud-vision-0.21.1/lib/google/cloud/vision/v1.rb:15:in `<top (required)>'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `require'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `block in require'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:240:in `load_dependency'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `require'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/google-cloud-vision-0.21.1/lib/google/cloud/vision/service.rb:18:in `<top (required)>'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `require'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `block in require'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:240:in `load_dependency'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `require'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/google-cloud-vision-0.21.1/lib/google/cloud/vision/project.rb:18:in `<top (required)>'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `require'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `block in require'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:240:in `load_dependency'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `require'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/google-cloud-vision-0.21.1/lib/google/cloud/vision.rb:17:in `<top (required)>'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `require'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `block in require'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:240:in `load_dependency'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `require'
	from /home/rails/app/services/cloud_vision.rb:5:in `<top (required)>'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `require'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `block in require'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:240:in `load_dependency'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:274:in `require'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:360:in `require_or_load'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:317:in `depend_on'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/dependencies.rb:233:in `require_dependency'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/railties-4.2.6/lib/rails/engine.rb:472:in `block (2 levels) in eager_load!'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/railties-4.2.6/lib/rails/engine.rb:471:in `each'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/railties-4.2.6/lib/rails/engine.rb:471:in `block in eager_load!'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/railties-4.2.6/lib/rails/engine.rb:469:in `each'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/railties-4.2.6/lib/rails/engine.rb:469:in `eager_load!'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/railties-4.2.6/lib/rails/engine.rb:346:in `eager_load!'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/railties-4.2.6/lib/rails/application/finisher.rb:56:in `each'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/railties-4.2.6/lib/rails/application/finisher.rb:56:in `block in <module:Finisher>'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/railties-4.2.6/lib/rails/initializable.rb:30:in `instance_exec'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/railties-4.2.6/lib/rails/initializable.rb:30:in `run'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/railties-4.2.6/lib/rails/initializable.rb:55:in `block in run_initializers'
	from /usr/local/rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/tsort.rb:228:in `block in tsort_each'
	from /usr/local/rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/tsort.rb:350:in `block (2 levels) in each_strongly_connected_component'
	from /usr/local/rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/tsort.rb:431:in `each_strongly_connected_component_from'
	from /usr/local/rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/tsort.rb:349:in `block in each_strongly_connected_component'
	from /usr/local/rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/tsort.rb:347:in `each'
	from /usr/local/rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/tsort.rb:347:in `call'
	from /usr/local/rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/tsort.rb:347:in `each_strongly_connected_component'
	from /usr/local/rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/tsort.rb:226:in `tsort_each'
	from /usr/local/rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/tsort.rb:205:in `tsort_each'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/railties-4.2.6/lib/rails/initializable.rb:54:in `run_initializers'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/railties-4.2.6/lib/rails/application.rb:352:in `initialize!'
	from /home/rails/config/environment.rb:5:in `<top (required)>'
	from /usr/local/rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from /usr/local/rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from config.ru:4:in `block in <main>'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/rack-1.6.4/lib/rack/builder.rb:55:in `instance_eval'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/rack-1.6.4/lib/rack/builder.rb:55:in `initialize'
	from config.ru:1:in `new'
	from config.ru:1:in `<main>'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/unicorn-5.1.0/lib/unicorn.rb:56:in `eval'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/unicorn-5.1.0/lib/unicorn.rb:56:in `block in builder'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/unicorn-5.1.0/lib/unicorn/http_server.rb:752:in `build_app!'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/unicorn-5.1.0/lib/unicorn/http_server.rb:617:in `init_worker_process'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/unicorn-5.1.0/lib/unicorn/http_server.rb:639:in `worker_loop'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/unicorn-5.1.0/lib/unicorn/http_server.rb:508:in `spawn_missing_workers'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/unicorn-5.1.0/lib/unicorn/http_server.rb:132:in `start'
	from /usr/local/rvm/gems/ruby-2.3.0/gems/unicorn-5.1.0/bin/unicorn:126:in `<top (required)>'
	from /usr/local/rvm/gems/ruby-2.3.0/bin/unicorn:23:in `load'
	from /usr/local/rvm/gems/ruby-2.3.0/bin/unicorn:23:in `<main>'
	from /usr/local/rvm/gems/ruby-2.0.0-p353/bin/ruby_executable_hooks:15:in `eval'
	from /usr/local/rvm/gems/ruby-2.0.0-p353/bin/ruby_executable_hooks:15:in `<main>'
```

**Comment by blowmage** [comment]
Can you do a `gem list` on your production environment? Were the gems installed successfully?

**Comment by jdraths** [comment]
**gem list**
```ruby
*** LOCAL GEMS ***

actionmailer (4.2.6)
actionpack (4.2.6)
actionview (4.2.6)
activejob (4.2.6)
activemodel (4.2.6)
activerecord (4.2.6)
activesupport (4.2.6)
activesupport-json_encoder (1.1.0)
addressable (2.4.0, 2.3.5)
ahoy_matey (1.4.0)
api-pagination (4.3.0)
arel (6.0.3)
atomic (1.1.14)
aws-sdk (2.2.31)
aws-sdk-core (2.2.31)
aws-sdk-resources (2.2.31)
aws-ses (0.5.0)
backup (4.2.3)
bcrypt (3.1.10)
bigdecimal (1.2.8)
bootstrap-chosen-rails (0.0.4)
bootstrap-sass (3.2.0.0)
bootstrap-will_paginate (0.0.10)
braintree (2.62.0)
browser (2.0.3)
buftok (0.2.0)
builder (3.2.2)
bundler (1.13.7, 1.11.2)
bundler-unload (1.0.2)
carrierwave (0.10.0)
carrierwave-aws (1.0.1)
CFPropertyList (2.3.2, 2.3.1)
chartkick (1.4.2)
chosen-rails (1.5.2)
chronic (0.10.2)
chunky_png (1.3.5)
client_side_validations (4.2.10)
coffee-rails (4.1.1)
coffee-script (2.4.1)
coffee-script-source (1.10.0)
commander (4.4.0)
compass (1.0.3)
compass-core (1.0.3)
compass-import-once (1.0.5)
compass-rails (3.0.2)
concurrent-ruby (1.0.1)
connection_pool (2.2.0)
descendants_tracker (0.0.3)
did_you_mean (1.0.0)
dogapi (1.11.0)
dropbox-sdk (1.5.1)
elasticsearch (2.0.0)
elasticsearch-api (2.0.0)
elasticsearch-transport (2.0.0)
equalizer (0.0.9)
errbase (0.0.3)
erubis (2.7.0)
ethon (0.9.1)
excon (0.49.0, 0.44.4)
execjs (2.6.0)
executable-hooks (1.3.2)
exif (1.0.1)
faraday (0.9.2, 0.8.8)
fcm (0.0.2)
ffi (1.9.10)
fission (0.5.0)
flowdock (0.4.0)
fog (1.37.0, 1.28.0)
fog-aliyun (0.1.0)
fog-atmos (0.1.0)
fog-aws (0.9.2, 0.1.1)
fog-brightbox (0.10.1, 0.7.1)
fog-core (1.37.0, 1.29.0)
fog-dynect (0.0.3)
fog-ecloud (0.3.0, 0.0.2)
fog-google (0.1.0)
fog-json (1.0.2, 1.0.0)
fog-local (0.3.0)
fog-powerdns (0.1.1)
fog-profitbricks (0.0.5, 0.0.2)
fog-radosgw (0.0.5, 0.0.3)
fog-riakcs (0.1.0)
fog-sakuracloud (1.7.5, 1.0.0)
fog-serverlove (0.1.2, 0.1.1)
fog-softlayer (1.1.0, 0.4.1)
fog-storm_on_demand (0.1.1, 0.1.0)
fog-terremark (0.1.0, 0.0.4)
fog-vmfusion (0.1.0, 0.0.1)
fog-voxel (0.1.0, 0.0.2)
fog-vsphere (0.6.3)
fog-xenserver (0.2.3)
fog-xml (0.1.2, 0.1.1)
formatador (0.2.5)
gem-wrappers (1.2.7)
geocoder (1.3.1)
get_process_mem (0.2.1)
globalid (0.3.6)
google-cloud-core (0.21.1)
google-cloud-vision (0.21.1)
google-gax (0.6.0)
google-protobuf (3.0.2 x86_64-linux)
googleapis-common-protos (1.3.4)
googleauth (0.5.1)
grim (1.2.0)
groupdate (2.5.2)
grpc (1.0.1 x86_64-linux)
hashie (3.4.6)
highline (1.7.8)
hipchat (1.0.1)
houston (2.2.3)
http (0.5.0)
http_parser.rb (0.6.0)
httparty (0.13.7, 0.12.0)
i18n (0.7.0)
image-picker-rails (0.2.4)
inflecto (0.0.2)
io-console (0.4.5)
ipaddress (0.8.3, 0.8.0)
jbuilder (2.4.1)
jmespath (1.2.2)
jquery-fileupload-rails (0.4.6)
jquery-rails (4.2.1, 4.1.1)
jquery-turbolinks (2.1.0)
jquery-ui-rails (5.0.5)
js_regex (1.1.0)
json (1.8.3, 1.8.2)
json_pure (1.8.3)
jwt (1.5.6)
kgio (2.10.0)
little-plugger (1.1.4)
logging (2.1.0)
loofah (2.0.3)
mail (2.6.4, 2.5.4)
memoist (0.15.0)
memoizable (0.4.0)
mime-types (3.0, 1.25.1)
mime-types-data (3.2016.0221)
mini_magick (4.4.0)
mini_portile (0.6.2)
mini_portile2 (2.0.0)
minitest (5.8.4, 5.8.3)
multi_json (1.11.2, 1.10.1)
multi_xml (0.5.5)
multipart-post (2.0.0, 1.2.0)
net-scp (1.2.1)
net-sftp (2.1.2)
net-ssh (2.9.2)
net-telnet (0.1.1)
nokogiri (1.6.7.2, 1.6.6.2)
oj (2.17.1)
open4 (1.3.0)
os (0.9.6)
pagerduty (2.0.0)
pdf-core (0.6.1)
pg (0.18.4)
polyglot (0.3.3)
posix-spawn (0.3.11)
power_assert (0.2.7, 0.2.6)
prawn (2.1.0)
prawn-table (0.2.2)
psych (2.0.17)
rack (1.6.4)
rack-attack (4.4.1)
rack-mini-profiler (0.10.1)
rack-protection (1.5.3)
rack-test (0.6.3)
rails (4.2.6)
rails-deprecated_sanitizer (1.0.3)
rails-dom-testing (1.0.7)
rails-html-sanitizer (1.0.3)
rails_12factor (0.0.3)
rails_serve_static_assets (0.0.5)
rails_stdout_logging (0.0.5)
railties (4.2.6)
raindrops (0.16.0)
rake (11.1.2, 10.4.2)
rb-fsevent (0.9.7)
rb-inotify (0.9.7)
rbvmomi (1.8.2)
rdoc (4.2.2, 4.2.1)
redis (3.2.2)
referer-parser (0.3.0)
regexp_parser (0.4.1)
request_store (1.3.0)
responders (2.1.2)
rly (0.2.3)
rollbar (2.11.5, 2.8.3, 2.7.1)
roo (2.3.2)
rubygems-bundler (1.4.4)
rubyzip (1.2.0)
rvm (1.11.3.9)
safely_block (0.1.0)
sass (3.4.22)
sass-rails (5.0.4)
searchkick (1.3.5)
sidekiq (4.1.1)
sidekiq-status (0.6.0)
signet (0.7.3)
simple_oauth (0.2.0)
sinatra (1.4.7)
skylight (0.10.5)
sprockets (3.5.2)
sprockets-rails (3.0.4)
test-unit (3.1.8, 3.1.5)
thor (0.19.1, 0.18.1)
thread_safe (0.3.5, 0.1.3)
tilt (2.0.2)
treetop (1.4.15)
trollop (2.1.2)
ttfunk (1.4.0)
turbolinks (2.5.3)
twitter (5.5.0)
typhoeus (1.1.0)
tzinfo (1.2.2)
uglifier (3.0.0)
unf (0.1.3)
unf_ext (0.0.6)
unicorn (5.1.0, 5.0.1)
unicorn-worker-killer (0.4.4)
user_agent_parser (2.3.0)
uuidtools (2.1.5)
whenever (0.9.4)
will_paginate (3.0.7)
xml-simple (1.1.5, 1.1.4)
```

**Comment by blowmage** [comment]
I've double and triple checked all the gems and I don't see anything in the dependencies that would allow this. All my local checks pass just fine. What are the differences in your development and production environments?

Do you also get this error when issuing `require "google/cloud/vision"` or `require "google/gax"` in an rails console in the production environment?

**Comment by jdraths** [comment]
Yea, I'm racking my brain trying to think of what the critical difference is.

In production `require "google/cloud/vision"` works, but `require "google/gax" #=> false`

**Comment by blowmage** [comment]
Calling `require "google/gax"` will return `true` if the library hasn't been loaded yet, and return `false` if it has already been required. The bad behavior that you opened this issue for was the require raising `LoadError`, which it doesn't look like the rails console does. That is very interesting...

**Comment by blowmage** [comment]
@jdraths To help you debug this further, I will need to ask you more detailed information about your production environment, and we may not want to disclose that information on this public issue. Please hit me up on Google Hangouts at mike@blowmage.com.

**Comment by blowmage** [comment]
I debugged with @jdraths offline and this behavior is specific to Unicorn on Nginx. The error occurs when Unicorn is started, not when an HTTP request is made. The library is available when running a rails console.

I suggested a work-around of moving the requires to the `initialize` method to get him unstuck, but I don't have a solution for why this happened in the first place.

**Comment by quartzmo** [comment]
Nice work @blowmage!

**Comment by blowmage** [comment]
@jdraths I am wondering if this might be an issue with using two different rubies in RVM. If you look at the backtrace in the issue description, the first two steps use ruby-2.0.0-p353, and the rest use ruby-2.3.0. Perhaps 

Are you using [RVM Wrappers](https://rvm.io/integration/init-d) in your unicorn configuration?

**Comment by jdraths** [comment]
@blowmage thanks for pointing that out. I had two rubies in my path. I removed the old one and now only have ruby-2.3.0... still have the same problem when the requires are moved outside of initialize. I'm digging into my unicorn setup to see if I have anything funky going on there.

**Comment by blowmage** [comment]
@jdraths What OS are you running? Ubuntu 14.04?

**Comment by jdraths** [comment]
yes, Ubuntu 14.04

**Comment by blowmage** [comment]
We have verified that googleapis/gax-ruby#54 is the cause for this, so I am going to close this issue. Thank you for opening the issue, helping diagnose the problem, and verifying the fix!

**Comment by blowmage** [comment]
This issue should be fixed with `google-cloud-datastore` version 0.24.0 that was released today. Please upgrade. Thanks for working with us on this!
